### PR TITLE
More generic and lightweight PyTorch integration

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This PR reworks the python bindings with the goal of being lighweight. 
Decisions and impl are backed by benchmarks on a strawman TcBuilder / TcFunction abstraction.

Running `python ./test_python/pytorch_example.py` prints something like:

```
#################################################################
raw unchecked_run naive options  Total CPU time to launch kernel: min 20us, p50 23us, p90 26us, max 2261us
raw unchecked_run naive options  Total CPU launch + GPU kernel time: min 431us, p50 435us, p90 441us, max 74723us
Tune with cache @ /tmp/8cca3161-0c96-4594-9c92-86707127eaec
Note that if you pass a fixed filename, you can reinforce an existing tuning state
Iteration 0     Jobs(Compiled, Evaluated)/total  (25, 25)/25   (best/median/worst)us: 321/721/17831
Iteration 1     Jobs(Compiled, Evaluated)/total  (25, 25)/25   (best/median/worst)us: 126/348/1513
Iteration 2     Jobs(Compiled, Evaluated)/total  (25, 25)/25   (best/median/worst)us: 126/345/555
#################################################################
raw unchecked_run tuned options  Total CPU time to launch kernel: min 19us, p50 25us, p90 28us, max 203us
raw unchecked_run tuned options  Total CPU launch + GPU kernel time: min 366us, p50 371us, p90 376us, max 605us
#################################################################
TcBuilder unchecked_run  Total CPU time to launch kernel: min 8us, p50 11us, p90 13us, max 1546us
TcBuilder unchecked_run  Total CPU launch + GPU kernel time: min 291us, p50 295us, p90 297us, max 1823us
#################################################################
TcFunction forward unchecked_run         Total CPU time to launch kernel: min 17us, p50 21us, p90 24us, max 498101us
TcFunction forward unchecked_run         Total CPU launch + GPU kernel time: min 299us, p50 303us, p90 307us, max 498359us
#################################################################
TcFunction backward unchecked_run        Total CPU time to launch kernel: min 174us, p50 246us, p90 399us, max 425us
TcFunction backward unchecked_run        Total CPU launch + GPU kernel time: min 1192us, p50 1282us, p90 1355us, max 1378us
#################################################################
MultiTcBuilder unchecked_run     Total CPU time to launch kernel: min 8us, p50 11us, p90 13us, max 1415us
MultiTcBuilder unchecked_run     Total CPU launch + GPU kernel time: min 292us, p50 295us, p90 297us, max 1691us
#################################################################
MultiTcFunction forward unchecked_run    Total CPU time to launch kernel: min 24us, p50 28us, p90 28us, max 498183us
MultiTcFunction forward unchecked_run    Total CPU launch + GPU kernel time: min 303us, p50 307us, p90 311us, max 498448us
#################################################################
MultiTcFunction backward unchecked_run   Total CPU time to launch kernel: min 212us, p50 348us, p90 402us, max 557us
MultiTcFunction backward unchecked_run   Total CPU launch + GPU kernel time: min 1242us, p50 1333us, p90 1356us, max 2025us
```

A large fraction of the overhead in backward seems to come from PyTorch autograd (around 200 microseconds for `matmul_grad`). The least intrusive way I found to benchmark this is to flick the `dump_backward_overhead` flag between the call into the lambda in `test_tc` and being in the `backward` function.

I wonder if this could come from saving the `TcBuilder` in `ctx` (but I doubt it).

Commit messages are still missing.

### CODE REVIEW

File: tc/aten/aten_compiler-inl.h [hunk 1]
```diff
@@ -82,5 +82,13 @@ void uncheckedRun(
   }
   return executor.uncheckedRun(rawInputs, rawOutputs);
 }
+
+template <typename Executor>
+void uncheckedRun(
```
Comment (ftynse) [inline tc/aten/aten_compiler-inl.h]: Why do we need this function?  It does not seem anyhow ATen-specific.

File: tc/aten/aten_compiler.h [hunk 2]
```diff
@@ -88,6 +88,15 @@ void uncheckedRun(
...
     std::vector<at::Tensor>& outputs);
+
+/// This is the "low-latency" mode in which we just propagate raw pointers
+/// Sizes are not checked and it is the user's responsibility to ensure that
+/// they match. If the user doesn't then segfault will likely occur.
+template <typename Executor>
+void uncheckedRun(
+    const Executor& executor,
+    const std::vector<const void*>& inputs,
+    std::vector<void*>& outputs);
```
Comment (ftynse) [inline tc/aten/aten_compiler.h]: The definition above has this argument as _const_ reference.  I'm surprised the compiler did not complain, maybe it never gets instantiated.

File: tensor_comprehensions/pybinds/tclib.cc [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+  py::list outputs;
+  for (auto& tensor : tensors) {
+    outputs.append(py::cast(torch::autograd::make_variable(tensor)));
+  }
+  return outputs;
+}
+
+} // namespace
+
+struct CudaCompilationUnit {
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: I'd rather get rid of this entire class.  This is again a cache/storage disguised as a compiler.  If you want to have a compilation cache, okay, but at least call it that way.

File: tensor_comprehensions/pybinds/tclib.cc [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+  CudaCompilationUnit(const std::string& def) : tc(def) {
+    initGlog();
+  }
+  std::string tc;
+  std::unordered_map<std::string, std::unique_ptr<tc::CudaTcExecutor>> compiled;
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: What is the key here?

File: tensor_comprehensions/pybinds/tclib.cc [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+  // the future. In any case, RAII and Python GC can just never work together.
+  void __enter__() const {
+    savedGenerations_ = tc::FLAGS_tuner_gen_generations;
+    savedPopulationSize_ = tc::FLAGS_tuner_gen_pop_size;
+    savedThreads_ = tc::FLAGS_tuner_threads;
+    savedDevices_ = tc::FLAGS_tuner_devices;
+    savedLogtostderr_ = FLAGS_logtostderr;
+    savedStderrthreshold_ = FLAGS_stderrthreshold;
+
+    tc::FLAGS_tuner_gen_generations = generations_;
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: Maybe we should stop relying on flags aka global variables for configuration.  Not in this PR, obviously.

File: tensor_comprehensions/pybinds/tclib.cc [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+          py::init<uint32_t, uint32_t, uint32_t, std::string, bool, uint32_t>(),
+          py::arg("generations") = tc::FLAGS_tuner_gen_generations,
+          py::arg("pop_size") = tc::FLAGS_tuner_gen_pop_size,
+          py::arg("threads") = tc::FLAGS_tuner_threads,
+          py::arg("devices") = tc::FLAGS_tuner_devices,
+          py::arg("logtostderr") = false,
+          // Suppress non-FATAL errors from the python user
+          py::arg("stderrthreshold") = google::ERROR);
+
+  py::class_<CudaTuner>(m, "Tuner", py::module_local())
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: Can't we have identical names for C++ and Python classes?

File: tensor_comprehensions/pybinds/tclib.cc [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+          })
+      .def(
+          "uncheckedRun",
+          [](CudaCompilationUnit& instance,
+             const std::string& entryPoint,
+             py::tuple& inputs,
+             py::tuple& outputs) {
+            CHECK_GE(outputs.size(), 1u)
+                << "uncheckedRun needs a tuple of output tensors to write into";
+            tc::aten::uncheckedRun(
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: Why can't you just call `instace.compiled.at(entryPoint)->uncheckedRun(...)` here?  There's no need for that free function in `tc::aten`.

**Comment by nicolasvasilache** [comment]
Ok this is getting closer. I still have a few things to tackle on Monday but I now see:

```
#################################################################
raw unchecked_run naive options  CPU allocation time: min 3us, p50 4us, p90 5us, max 1212us
raw unchecked_run naive options  Total CPU time to launch kernel: min 14us, p50 16us, p90 18us, max 1529us
raw unchecked_run naive options  Total CPU launch + GPU kernel time: min 432us, p50 436us, p90 439us, max 62467us
Tune with cache @ /tmp/f18491f9-420c-4c09-ac38-fe7cb86af73e
Note that if you pass a fixed filename, you can reinforce an existing tuning state
Iteration 0     Jobs(Compiled, Evaluated)/total  (25, 25)/25   (best/median/worst)us: 349/1512/50743
#################################################################
raw unchecked_run tuned options  CPU allocation time: min 6us, p50 11us, p90 13us, max 83us
raw unchecked_run tuned options  Total CPU time to launch kernel: min 23us, p50 37us, p90 43us, max 284us
raw unchecked_run tuned options  Total CPU launch + GPU kernel time: min 442us, p50 455us, p90 463us, max 713us
#################################################################
TcBuilder unchecked_run  CPU allocation time: min 4us, p50 5us, p90 6us, max 966us
TcBuilder unchecked_run  Total CPU time to launch kernel: min 15us, p50 19us, p90 23us, max 1260us
TcBuilder unchecked_run  Total CPU launch + GPU kernel time: min 219us, p50 277us, p90 281us, max 1516us
#################################################################
Functional forward unchecked run         CPU allocation time: min 5us, p50 5us, p90 7us, max 816us
Functional forward unchecked run         Total CPU time to launch kernel: min 36us, p50 39us, p90 42us, max 604671us
Functional forward unchecked run         Total CPU launch + GPU kernel time: min 289us, p50 292us, p90 297us, max 604892us
#################################################################
Functional backward unchecked run        CPU allocation time: min 0us, p50 1us, p90 2us, max 5us
Functional backward unchecked run        Total CPU time to launch kernel: min 1579us, p50 1998us, p90 2769us, max 4974us
Functional backward unchecked run        Total CPU launch + GPU kernel time: min 2686us, p50 3061us, p90 3917us, max 5243us
```

The backward CPU overhead is prob. related to that synchronization bug.
The rest looks reasonable to me, maybe we could squeeze out a few more micros but I'd like to see concrete cases where this is an issue.

Also, cc @zdevito

File: test_python/pytorch_example.py [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+        "matmul",
+        (mat1, mat2))
+
+################################################################################
+# 4. Simple torch.autograd.Function backed by TcBuilder
+################################################################################
+class TcFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, tc_builder, *inputs):
+        ctx.tc_builder = tc_builder
```
Comment (ftynse) [inline test_python/pytorch_example.py]: I'm not sure what kind of object `ctx` is and if you can just add fields to it.  Pytorch documentation is quite scarce about this, saying "you can use save_for_backward" to stash any information for the backward path.  I suppose you know what you are doing here.

File: test_python/pytorch_example.py [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+# 1. Use the C++ API to build a low-overhead compilation cache and time it
+################################################################################
+from tensor_comprehensions.tclib import CompilationCache
+
+compilation_cache = CompilationCache(mm)
+# Compilation returns an allocated tuple of outputs with the proper shapes.
+# Allocation overhead is negligible compared to compilation overhead.
+compilation_cache.compile("matmul", (mat1, mat2), MappingOptions())
+# unchecked_run on  tensors
+time_tc(100,
```
Comment (ftynse) [inline test_python/pytorch_example.py]: would be nice to have an example without timing

File: test_python/pytorch_example.py [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+}
+"""
+mat1, mat2 = torch.randn(300, 400).cuda(), torch.randn(400, 500).cuda()
+
+################################################################################
+# 1. Use the C++ API to build a low-overhead compilation cache and time it
+################################################################################
+from tensor_comprehensions.tclib import CompilationCache
+
+compilation_cache = CompilationCache(mm)
```
Comment (ftynse) [inline test_python/pytorch_example.py]: can we make a minimal example without compilation cache?

File: tensor_comprehensions/pybinds/tclib.cc [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+    generations_ = generations;
+    populationSize_ = populationSize;
+    threads_ = threads;
+    devices_ = devices;
+    logtostderr_ = logtostderr;
+    stderrthreshold_ = stderrthreshold;
+  }
+  // __enter__ / __exit__ in case we want to use a ContextManager in Python in
+  // the future. In any case, RAII and Python GC can just never work together.
+  void __enter__() const {
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: do we have to use these names in C++ to make it work?

File: tensor_comprehensions/pybinds/tclib.cc [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+ * We want this to be lightweight so we cannot afford:
+ *   1. python-side dictionary manipulations,
+ *   2. parsing the TC string more than necessary (compilation and new
+ *      allocations are acceptable)
+ *   3. large string hashing
+ *
+ * Creating a key must be cheap, we cannot afford multiple conversions on the
+ * way from ATen tensors to TensorInfo.
+ */
+struct CompilationCache {
```
Comment (ftynse) [inline tensor_comprehensions/pybinds/tclib.cc]: I'm not a fa of providing extra functionality in the bindings only, but this is strictly better that what we had before

File: test_python/pytorch_example.py (hunk 8 above)
Comment (nicolasvasilache) [reply test_python/pytorch_example.py]: yeah that's a PyTorch-ism

File: test_python/pytorch_example.py (hunk 10 above)
Comment (nicolasvasilache) [reply test_python/pytorch_example.py]: The idea in this file is to use the python bindings to build python abstractions.
But this should not be the entry point for a user.
I'll rename the file and make it clear that this is internal.
The next PR will reimplement the user-facing API with these primitives.

Comment (ftynse) [reply test_python/pytorch_example.py]: yes, the filename got me confused, it implies the the code is an example of using tc in pytorch

**Pull request merged**